src/components/encoding-pane/property-editor-schema.ts

```diff
@@ -0,0 +1,195 @@
+import {Channel} from 'vega-lite/build/src/channel';
+import * as vlSchema from 'vega-lite/build/vega-lite-schema.json';
+import {ShelfFieldDef, ShelfId} from '../../models/shelf/spec';
+
+export interface SchemaProperties {
+  [key: string]: SchemaProperty;
+}
+
+export interface ObjectSchema {
+  type: 'object';
+  title?: string;
+  properties: SchemaProperties;
+}
+export interface StringSchema {
+  type: 'string';
+  title: string;
+  enum?: string[];
+}
+
+export function isStringSchema(schema: SchemaProperty): schema is StringSchema {
+  return schema.type === 'string';
+}
+
+export type SchemaProperty = ObjectSchema | StringSchema;
+
+export interface UISchema {
+  [key: string]: UISchemaItem;
+}
+
+// NOTE: keys for these interfaces follow the requirement for react-jsonSchema form
+// (https://mozilla-services.github.io/react-jsonschema-form/)
+export interface UISchemaItem {
+  'ui:autofocus'?: boolean;
+  'ui:widget'?: string;
+  'ui:placeholder'?: string;
+  'ui:emptyValue'?: string;
+}
+
+export interface PropertyEditorSchema {
+  uiSchema: UISchema;
+  schema: ObjectSchema;
+}
+
+const DEFAULT_TEXT_UISCHEMA: UISchemaItem = {
+  'ui:autofocus': true,
+  'ui:emptyValue': ''
+};
+
+const DEFAULT_SELECT_UISCHEMA: UISchemaItem = {
+  'ui:widget': 'select',
+  'ui:placeholder': 'auto',
+  'ui:emptyValue': 'auto'
+};
+
+const POSITION_FIELD_QUANTITATIVE_INDEX = {
+  'Common': [
+    {
+      prop: 'scale',
+      nestedProp: 'type'
+    },
+    {
+      prop: 'axis',
+      nestedProp: 'title'
+    },
+    {
+      prop: 'stack'
+    }
+  ],
+  'Scale': ['type'].map(p => ({prop: 'scale', nestedProp: p})),
+  'Axis': ['orient', 'title'].map(p => ({prop: 'axis', nestedProp: p}))
+};
+
+const POSITION_FIELD_NOMINAL_INDEX = {
+  'Scale': ['type'].map(p => ({prop: 'scale', nestedProp: p})),
+  'Axis': ['orient', 'title'].map(p => ({prop: 'axis', nestedProp: p}))
+};
+
+const POSITION_FIELD_TEMPORAL_INDEX = {
+  'Scale': ['type'].map(p => ({prop: 'scale', nestedProp: p})),
+  'Axis': ['orient', 'title'].map(p => ({prop: 'axis', nestedProp: p}))
+};
+
+// Capitalize first letter for aesthetic purposes in form
+function toTitleCase(str: string) {
+  return str.replace(/\w\S*/g, txt => {
+    return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();
+  });
+}
+
+function generateTitle(prop: string, nestedProp: string, propTab: string): string {
+  let title;
+  if (propTab === 'Common') {
+    title = nestedProp ? prop + ' ' + nestedProp : prop;
+  } else {
+    title = nestedProp || prop;
+  }
+
+  return toTitleCase(title);
+}
+
+export function getPropertyEditorSchema(prop: string, nestedProp: string, propTab: string): PropertyEditorSchema {
+  const title = generateTitle(prop, nestedProp, propTab);
+  const baseSchemaProp = {
+    title: title
+  };
+  if (prop === 'scale') {
+    if (nestedProp === 'type') {
+      const schemaProperty: StringSchema = {
+        ...baseSchemaProp as StringSchema,
```

Comment:
Why using `as`?  looks like a hack. 

Why having baseSchemaProp at all anyway?   (Why not adding `title` directly to each of these cases?)